madara/crates/client/db/src/watch.rs

```
@@ -0,0 +1,83 @@
+use crate::{MadaraBackend, MadaraStorageError};
+use mp_block::{header::PendingHeader, MadaraBlockInfo, MadaraPendingBlockInfo};
+use std::sync::Arc;
+use tokio::sync::{broadcast, watch};
+
+pub type ClosedBlocksReceiver = broadcast::Receiver<Arc<MadaraBlockInfo>>;
+pub type PendingBlockReceiver = watch::Receiver<Arc<MadaraPendingBlockInfo>>;
+
+fn make_fake_pending_block(parent_block: Option<&MadaraBlockInfo>) -> Arc<MadaraPendingBlockInfo> {
+    let Some(parent_block) = parent_block else {
+        return Default::default(); // No genesis block, we have make it all up
+    };
+    MadaraPendingBlockInfo {
+        header: PendingHeader {
+            parent_block_hash: parent_block.block_hash,
+            sequencer_address: parent_block.header.sequencer_address,
+            block_timestamp: parent_block.header.block_timestamp, // Junk timestamp: unix epoch
+            protocol_version: parent_block.header.protocol_version,
+            l1_gas_price: parent_block.header.l1_gas_price.clone(),
+            l1_da_mode: parent_block.header.l1_da_mode,
+        },
+        tx_hashes: vec![],
+    }
+    .into()
+}
+
+pub(crate) struct BlockWatch {
```

So if I understand this correctly we use this struct to replace the _many_ calls we did to the pending block in db with a _single_ call on pending update, the result of which we store into a `tokio::sync::watch` channel. If no value is present we default to dummy pending block. Is that correct?

---

yes, and we can watch for changes in the pending block too
although i don't see i do it here